proxy/src/connection.rs

```diff
@@ -254,6 +250,39 @@ impl Future for ConditionallyUpgradeServerToTls {
     }
 }
 
+impl ConditionallyUpgradeServerToTlsInner {
+    /// Polls the underlying socket for more data and buffers it.
+    ///
+    /// The buffer is matched for a TLS client hello message.
+    ///
+    /// `None` is returned if the underlying socket has closed.
+    fn poll_match_client_hello(&mut self) -> Poll<Option<tls::conditional_accept::Match>, io::Error> {
+        let sz = try_ready!(self.socket.read_buf(&mut self.peek_buf));
+        if sz == 0 {
+            return Ok(None.into())
```

Comment:
With the current logic in the caller, returning `None` is the same as returning `NotMatched` so you might as well return `NotMatched`; as things currently are, there's no need for using `Option` in the return type of this function.

Now, the question is whether this is actually the behavior we want. Basically in this situation we have an `Incomplete` state. If we aren't even sure if it looks like a TLS handshake then returning `NotMatched` definitely makes sense since it might be some weird (binary) protocol that just happens to have a message that looks like a TLS record header. If we were more certain that it is TLS then we should return `Matched` and then let the TLS machinery eventually abort the connection due to the truncated ClientHello. However, right now conditional_accept.rs doesn't clearly distinguish between those two cases and I don't think we should block this on adding such fine-grained (and mostly heuristic) determination to conditional_accept.rs, so I think we can just return `NotMatched` unconditionally here for now. However, we should add a comment here like this:

```rust
// XXX: It is ambiguous whether this is the start of a TLS handshake or not.
// For now, resolve the ambiguity in favor of plaintext. TODO: revisit this
// when we add support for TLS policy.
```